### PR TITLE
Add weightlessness status effect.

### PR DESCRIPTION
**Pull request by F77F**
Fixes https://github.com/space-wizards/space-station-14/issues/2378
placeholder texture, obv

image

### CODE REVIEW

**Review by metalgearsloth — COMMENTED** [review]
I'm not a huge fan of calling this every tick as it seems wasteful IMO, especially when IsWeightless isn't the cheapest thing in the world.

**Comment by ShadowCommander** [comment]
Huh, it's more complicated than I thought, but yeah it should probably be done through an event.

**Comment by F77F** [comment]
yeah, i figured. i assumed an event  like 'OnWeightlessChanged' already  existed or something, but it didn't.

File: Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs [hunk 1]
```diff
@@ -14,12 +15,15 @@ public sealed class MovementIgnoreGravityComponent : Component
 
     public static class GravityExtensions
     {
+        public static Action<IEntity,bool> ?OnWeightlessChanged;
```
Comment (PaulRitter) [inline Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:18]: nitpick, but looks good imo.

dunno how i feel about the global event, that might be costy real quick. might be better to add a local event somewhere on the entity?
```suggestion
        public static Action<IEntity,bool>? OnWeightlessChanged;
```

**Review by PaulRitter — APPROVED** [review]

File: Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs (hunk 1 above)
Comment (F77F) [reply Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:18]: > somewhere on the entity?

To the IEntity interface? if so, i did consider it for a minute.

Comment (PaulRitter) [reply Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:18]: no never ever do something like that on an entity.

if anything the options would be
a) find a component that relates to weighlessness and add it on there
b) create a new component to hold the event

Comment (PaulRitter) [reply Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:18]: hold on i got an idea

File: Content.Shared/GameObjects/Components/Mobs/SharedWeightlessStatusComponent.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            GravityExtensions.OnWeightlessChanged += (entity,isWeightless) =>
+            {
+                if(entity != Owner) return;
+                UpdateStatus(isWeightless);
+            };
+        }
```
Comment (PaulRitter) [inline Content.Shared/GameObjects/Components/Mobs/SharedWeightlessStatusComponent.cs:20]:
```suggestion
```

File: Content.Shared/GameObjects/Components/Mobs/SharedWeightlessStatusComponent.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+        protected void UpdateStatus(bool isWeightless)
```
Comment (PaulRitter) [inline Content.Shared/GameObjects/Components/Mobs/SharedWeightlessStatusComponent.cs:22]:
```suggestion
        public void UpdateStatus(bool isWeightless)
```

File: Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs [hunk 4]
```diff
@@ -14,12 +15,15 @@ public sealed class MovementIgnoreGravityComponent : Component
...
         public static bool IsWeightless(this IEntity entity, IPhysicsManager? physicsManager = null)
         {
             physicsManager ??= IoCManager.Resolve<IPhysicsManager>();
 
-            return !entity.HasComponent<MovementIgnoreGravityComponent>() &&
+            bool isWeightless = !entity.HasComponent<MovementIgnoreGravityComponent>() &&
                    physicsManager.IsWeightless(entity.Transform.Coordinates);
+            OnWeightlessChanged?.Invoke(entity,isWeightless);
```
Comment (PaulRitter) [inline Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:25]:
```suggestion
            if(entity.TryGetComponent<SharedWeightlessStatusComponent> statusComp)
                statusComp.UpdateStatus(isWeightless);
```

File: Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs (hunk 1 above)
Comment (PaulRitter) [reply Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:18]: also just realized you are using `Action<IEntity,bool>?`. this means you can only subscribe one (1) method to this "event"

Comment (PaulRitter) [inline Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:18]: if you are using my other suggestion, this should be removed
```suggestion
```

**Review by PaulRitter — CHANGES_REQUESTED** [review]
this *should* work. cleaner imo

**Comment by F77F** [comment]
@PaulRitter i implemented your changes and am getting a key not found error when the client connects to server and presumably tries to handle the weightlesscomponent 

```Exception has occurred: CLR/System.Collections.Generic.KeyNotFoundException
An unhandled exception of type 'System.Collections.Generic.KeyNotFoundException' occurred in System.Private.CoreLib.dll: 'The given key 'Content.Shared.GameObjects.Components.Mobs.SharedWeightlessStatusComponent' was not present in the dictionary.'
   at System.ThrowHelper.ThrowKeyNotFoundException[T](T key)
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Robust.Shared.GameObjects.ComponentManager.TryGetComponent(EntityUid uid, Type type, IComponent& component) in E:\edev\ss14\RobustToolbox\Robust.Shared\GameObjects\ComponentManager.cs:line 397
   at Robust.Shared.GameObjects.ComponentManager.TryGetComponent[T](EntityUid uid, T& component) in E:\edev\ss14\RobustToolbox\Robust.Shared\GameObjects\ComponentManager.cs:line 381
   at Robust.Shared.GameObjects.Entity.TryGetComponent[T](T& component) in E:\edev\ss14\RobustToolbox\Robust.Shared\GameObjects\Entity.cs:line 298
   at Content.Shared.GameObjects.Components.Movement.GravityExtensions.IsWeightless(IEntity entity, IPhysicsManager physicsManager) in E:\edev\ss14\Content.Shared\GameObjects\Components\Movement\MovementIgnoreGravityComponent.cs:line 25
   at Content.Shared.GameObjects.EntitySystems.SharedMoverSystem.UpdateKinematics(ITransformComponent transform, IMoverComponent mover, IPhysicsComponent physics) in E:\edev\ss14\Content.Shared\GameObjects\EntitySystems\SharedMoverSystem.cs:line 59
   at Content.Client.GameObjects.EntitySystems.MoverSystem.FrameUpdate(Single frameTime) in E:\edev\ss14\Content.Client\GameObjects\EntitySystems\MoverSystem.cs:line 35
   at Content.Client.GameObjects.EntitySystems.MoverSystem.Update(Single frameTime) in E:\edev\ss14\Content.Client\GameObjects\EntitySystems\MoverSystem.cs:line 40
   at Robust.Shared.GameObjects.EntitySystemManager.Update(Single frameTime) in E:\edev\ss14\RobustToolbox\Robust.Shared\GameObjects\EntitySystemManager.cs:line 250
   at Robust.Client.GameStates.ClientGameStateManager.ApplyGameState() in E:\edev\ss14\RobustToolbox\Robust.Client\GameStates\ClientGameStateManager.cs:line 301
   at Robust.Client.GameController.Tick(FrameEventArgs frameEventArgs) in E:\edev\ss14\RobustToolbox\Robust.Client\GameController.cs:line 272
   at Robust.Client.GameController.<MainLoop>b__65_0(Object sender, FrameEventArgs args) in E:\edev\ss14\RobustToolbox\Robust.Client\GameController\GameController.Standalone.cs:line 84
   at Robust.Shared.Timing.GameLoop.Run() in E:\edev\ss14\RobustToolbox\Robust.Shared\Timing\GameLoop.cs:line 200
   at Robust.Client.GameController.MainLoop(DisplayMode mode) in E:\edev\ss14\RobustToolbox\Robust.Client\GameController\GameController.Standalone.cs:line 113
   at Robust.Client.GameController.ParsedMain(CommandLineArgs args, Boolean contentStart) in E:\edev\ss14\RobustToolbox\Robust.Client\GameController\GameController.Standalone.cs:line 59
   at Robust.Client.GameController.Start(String[] args, Boolean contentStart) in E:\edev\ss14\RobustToolbox\Robust.Client\GameController\GameController.Standalone.cs:line 35
   at Robust.Client.GameController.Main(String[] args) in E:\edev\ss14\RobustToolbox\Robust.Client\GameController\GameController.Standalone.cs:line 21```

**Comment by Zumorica** [comment]
Maybe due to a missing `[Serializable, NetSerializable]` somewhere?

**Comment by metalgearsloth** [comment]
You used a ComponentReference if you're trying to get the shared one right?

**Comment by F77F** [comment]
> You used a ComponentReference if you're trying to get the shared one right?

in the derived content.server class? yes.

**Comment by metalgearsloth** [comment]
And the shared class is abstract?

**Comment by F77F** [comment]
> And the shared class is abstract?

```cs
namespace Content.Shared.GameObjects.Components.Mobs
{
    public abstract class SharedWeightlessStatusComponent : Component
    {

**Comment by metalgearsloth** [comment]
I just had a look and there's no component on the client, only on the server.

**Comment by F77F** [comment]
> I just had a look and there's no component on the client, only on the server.

Yeah i tried adding it to the ignore list, and also tried making a clientside one and it still threw the error. I'll try again though.

**Comment by metalgearsloth** [comment]
You'll need to make a client-side one and componentreference shared for it to not throw.

**Comment by F77F** [comment]
> You'll need to make a client-side one and componentreference shared for it to not throw.

alright thankya

**Comment by metalgearsloth** [comment]
Looking at what you've done so far it'd be better if instead of calling UpdateStatus directly it either sent a message over the EventBus or called the entity's SendMessage method (probably the former?) so that way if anything else also needed to check for Weightlessness they could do it.

This would also mean you could just make a system for weightlessness that subscribes to the message and updates the entity's StatusEffect is applicable which would mean less memory usage (fewer components in use).

**Comment by F77F** [comment]
well, it works and i dont really feel like doing all that.

**Comment by F77F** [comment]
nah just kidding lol

**Comment by F77F** [comment]
> message over the EventBu

how will i do that from Content.Shared ?

**Comment by metalgearsloth** [comment]
If you already have the entity you can do like Owner.EntityManager.EventBus.RaiseEvent(EventSource.Local, new WeightlessChangeMessage(value)) or the likes (you'd need to make a WeightlessChangeMessage : EntitySystemMessage)

Then on the system you could do SubscribeLocalEvent<WeightlessChangeMessage>(Method that handles it here) and just have that check if the entity has a status effect and if it does update it.

**Comment by F77F** [comment]
@metalgearsloth thank you for the guidance on this!

File: Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs [hunk 5]
```diff
@@ -18,8 +21,10 @@ public static bool IsWeightless(this IEntity entity, IPhysicsManager? physicsMan
         {
             physicsManager ??= IoCManager.Resolve<IPhysicsManager>();
 
-            return !entity.HasComponent<MovementIgnoreGravityComponent>() &&
+            bool isWeightless = !entity.HasComponent<MovementIgnoreGravityComponent>() &&
                    physicsManager.IsWeightless(entity.Transform.Coordinates);
+            entity.EntityManager.EventBus.RaiseEvent<WeightlessChangeMessage>(EventSource.Local, new WeightlessChangeMessage(entity,isWeightless));
```
Comment (metalgearsloth) [inline Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:26]: I don't think you need the generic's type here?

File: Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs [hunk 6]
```diff
@@ -18,8 +21,10 @@ public static bool IsWeightless(this IEntity entity, IPhysicsManager? physicsMan
         {
             physicsManager ??= IoCManager.Resolve<IPhysicsManager>();
 
-            return !entity.HasComponent<MovementIgnoreGravityComponent>() &&
+            bool isWeightless = !entity.HasComponent<MovementIgnoreGravityComponent>() &&
```
Comment (metalgearsloth) [inline Content.Shared/GameObjects/Components/Movement/MovementIgnoreGravityComponent.cs:24]: var IMO

File: Content.Shared/GameObjects/EntitySystemMessages/WeightlessChangeMessage.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.GameObjects.EntitySystemMessages
+{
+    [Serializable, NetSerializable]
+    public sealed class WeightlessChangeMessage : EntitySystemMessage
+    {
+        public readonly IEntity Entity;
+        public readonly bool IsNowWeightless;
```
Comment (metalgearsloth) [inline Content.Shared/GameObjects/EntitySystemMessages/WeightlessChangeMessage.cs:12]: Could just call it like "Weightless" but this is pedantic so you can ignore it if you want.

File: Content.Server/GameObjects/EntitySystems/WeightlessStatusSystem.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
+using Content.Shared.GameObjects.Components.Mobs;
+using Content.Shared.GameObjects.EntitySystemMessages;
+using Robust.Shared.GameObjects.Systems;
+
+namespace Content.Shared.GameObjects.EntitySystems
+{
+
+    public class WeightlessStatusSystem : EntitySystem
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/EntitySystems/WeightlessStatusSystem.cs:8]: Could make it sealed given its doubtful anything would inherit it.

File: Content.Client/GameObjects/Components/Mobs/WeightlessStatusComponent.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+using Content.Shared.GameObjects.Components.Mobs;
+using Robust.Shared.GameObjects;
+
+
+[RegisterComponent]
+[ComponentReference(typeof(SharedWeightlessStatusComponent))]
+public class WeightlessStatusComponent : SharedWeightlessStatusComponent
+{
+
+}
```
Comment (metalgearsloth) [inline Content.Client/GameObjects/Components/Mobs/WeightlessStatusComponent.cs:10]: Don't think you'll need the components anymore.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Rest LGTM if it works in game, once changes made ship it.

**Comment by F77F** [comment]
@metalgearsloth tell me what you think now. thankya again

**Comment by metalgearsloth** [comment]
> 
> 
> @metalgearsloth tell me what you think now. thankya again

Looks mostly good, still has the sharedweightless component which can be removed (same with on the entity), rest looks good!

**Comment by F77F** [comment]
> > @metalgearsloth tell me what you think now. thankya again
> 
> Looks mostly good, still has the sharedweightless component which can be removed (same with on the entity), rest looks good!

oh damn. i keep thinking i need the components 🤦

**Comment by F77F** [comment]
also @metalgearsloth. what do you mean by 
> (same with on the entity),

?

**Comment by metalgearsloth** [comment]
Under human.yml it still has the reference to WeightlessStatus

**Comment by F77F** [comment]
gotchya

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**